core/server/master/src/main/java/alluxio/master/metrics/MetricsStore.java

```diff
@@ -145,21 +140,30 @@ private static void putReportedMetrics(IndexedSet<Metric> metricSet, String inst
         // FULL_NAME_INDEX is a unique index, so getFirstByField will return the same results as
         // getByField
         Metric oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        double oldVal = oldMetric == null ? 0.0 : oldMetric.getValue();
-        Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
-            metric.getMetricType(), metric.getName(), oldVal + metric.getValue());
-        for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
-          newMetric.addTag(tag.getKey(), tag.getValue());
+        if (oldMetric != null) {
+          metric.addValue(metric.getValue());
+        } else {
+          Metric newMetric = new Metric(metric.getInstanceType(), metric.getHostname(),
+              metric.getMetricType(), metric.getName(), metric.getValue());
+          for (Map.Entry<String, String> tag : metric.getTags().entrySet()) {
+            newMetric.addTag(tag.getKey(), tag.getValue());
+          }
+          synchronized (metricSet) {
+            oldMetric = metricSet.getFirstByField(FULL_NAME_INDEX, metric.getFullMetricName());
+            if (oldMetric != null) {
+              metric.addValue(metric.getValue());
+            } else {
+              metricSet.add(newMetric);
+            }
+          }
         }
-        metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        newMetrics.add(newMetric);
       } else {
-        metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
-        newMetrics.add(metric);
+        synchronized (metricSet) {
+          metricSet.removeByField(FULL_NAME_INDEX, metric.getFullMetricName());
```

Comment:
Do you know why the metrics are updated in this way (delete all metrics, then add all new metrics)? It seems inefficient, and we should probably be able to update existing data structures in an atomic way?